src/App.tsx

```diff
@@ -1,14 +1,53 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { Todo } from './types/Todo';
+import { getTodos } from './api';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [query, setQuery] = useState('');
+  const [filter, setFilter] = useState('');
+
+  const getVisibleTodos = (
+    allTodos: Todo[],
+    searchQuery: string,
+    selectedFilter: string,
+  ) => {
+    const lowerCase = (text: string) => text.toLowerCase().trim();
+    let visibleTodos = [...allTodos];
+
+    if (query) {
+      visibleTodos = visibleTodos
+        .filter(todo => lowerCase(todo.title).includes(lowerCase(searchQuery)));
+    }
+
+    return visibleTodos
+      .filter(todo => {
+        switch (selectedFilter) {
+          case 'completed':
+            return todo.completed;
+          case 'active':
+            return !todo.completed;
+          default:
+            return visibleTodos;
+        }
+      });
+  };
```

Comment:
Take it outside the component and wrap the function call in a useMemo, where to provide a filter, query dependency